File: tools/shot.py
#!/usr/bin/env python3
"""Headless TUI screenshots: run a binary in a private tmux server, drive it with keys and
mouse, capture with colours and rasterize to PNG (or print plain text).

Examples
  tools/shot.py -- ./target/debug/showcase --page controls
  tools/shot.py -k "Tab Tab Enter" -o /tmp/a.png -- ./target/debug/showcase --page inputs
  tools/shot.py -k "Down Down" --mouse "click 40 12" --text -- ./target/debug/showcase
  tools/shot.py --script 'Tab;shot /tmp/1.png;Enter;wait 0.5;shot /tmp/2.png' -- ./target/debug/showcase

Key names are tmux send-keys names (Tab, BTab, Enter, Escape, Up, Down, Left, Right, Space,
PageUp, PageDown, Home, End, F1.., C-p (ctrl), M-1 (alt), plain characters). `Escape` is
always sent alone with a pause (crossterm would otherwise read ESC+key as Alt+key).

Mouse steps (SGR sequences): `click X Y`, `down X Y`, `up X Y`, `drag X Y`, `move X Y`,
`wheelup X Y`, `wheeldown X Y` — 0-based cell coordinates.
"""
import argparse
import itertools
import os
import re
import subprocess
import sys
import time

FONT_CANDIDATES = [
    "/usr/share/fonts/TTF/JetBrainsMonoNerdFontMono-Regular.ttf",
    "/usr/share/fonts/TTF/JetBrainsMono-Regular.ttf",
    "/usr/share/fonts/TTF/DejaVuSansMono.ttf",
    "/usr/share/fonts/truetype/dejavu/DejaVuSansMono.ttf",
]
FONT_BOLD = [
    "/usr/share/fonts/TTF/JetBrainsMonoNerdFontMono-Bold.ttf",
    "/usr/share/fonts/TTF/JetBrainsMono-Bold.ttf",
    "/usr/share/fonts/TTF/DejaVuSansMono-Bold.ttf",
    "/usr/share/fonts/truetype/dejavu/DejaVuSansMono-Bold.ttf",
]


_COUNTER = itertools.count()


class Tmux:
    def __init__(self, cmd, cols, rows):
        self.sock = f"shot-{os.getpid()}-{next(_COUNTER)}"
        self.t = ["tmux", "-L", self.sock]
        shell = " ".join(_q(c) for c in cmd) + " 2>/tmp/shot-%d.err" % os.getpid()
        # Agent shells often export NO_COLOR=1 / TERM=dumb: crossterm would then strip every
        # colour. The tmux server (and thus the app) inherits this environment, so scrub it.
        env = {k: v for k, v in os.environ.items() if k not in ("NO_COLOR", "TMUX", "TMUX_PANE")}
        env["TERM"] = "xterm-256color"
        env["COLORTERM"] = "truecolor"
        subprocess.run([*self.t, "new-session", "-d", "-s", "s", "-x", str(cols), "-y", str(rows), shell], check=True, env=env)
        subprocess.run([*self.t, "set", "-g", "status", "off"], check=False)
        self.cols, self.rows = cols, rows

    def keys(self, *ks, wait=0.25):
        for k in ks:
            subprocess.run([*self.t, "send-keys", "-t", "s", k], check=True)
            if k in ("Escape", "Esc"):
                time.sleep(0.35)
        time.sleep(wait)

    def mouse(self, kind, x, y, wait=0.15):
        codes = {"down": (0, "M"), "up": (0, "m"), "drag": (32, "M"), "move": (35, "M"), "wheelup": (64, "M"), "wheeldown": (65, "M")}
        if kind == "click":
            self.mouse("down", x, y, 0.05)
            self.mouse("up", x, y, wait)
            return
        b, suffix = codes[kind]
        seq = f"\x1b[<{b};{x + 1};{y + 1}{suffix}"
        subprocess.run([*self.t, "send-keys", "-t", "s", "-l", seq], check=True)
        time.sleep(wait)

    def capture(self, ansi=True):
        args = [*self.t, "capture-pane", "-t", "s", "-p"] + (["-e"] if ansi else [])
        return subprocess.run(args, capture_output=True, text=True, check=True).stdout

    def alive(self):
        r = subprocess.run([*self.t, "list-panes", "-t", "s", "-F", "#{pane_dead}"], capture_output=True, text=True)
        return r.returncode == 0 and r.stdout.strip() == "0"

    def close(self):
        subprocess.run([*self.t, "kill-server"], capture_output=True)
        err = f"/tmp/shot-{os.getpid()}.err"
        out = open(err).read() if os.path.exists(err) else ""
        if os.path.exists(err):
            os.remove(err)
        return out


def _q(s):
    return "'" + s.replace("'", "'\\''") + "'"


def render(ansi_text, cols, rows, out_path):
    try:
        from PIL import Image, ImageDraw, ImageFont
        import unicodedata
    except ImportError:
        sys.exit("PIL not installed: pip install pillow (or use --text)")
    font = next((ImageFont.truetype(p, 18) for p in FONT_CANDIDATES if os.path.exists(p)), ImageFont.load_default())
    fontb = next((ImageFont.truetype(p, 18) for p in FONT_BOLD if os.path.exists(p)), font)
    cw, ch = 11, 22
    img = Image.new("RGB", (cols * cw, rows * ch), (0, 0, 0))
    d = ImageDraw.Draw(img)

    def c256(n):
        if n < 16:
            base = [(0, 0, 0), (205, 0, 0), (0, 205, 0), (205, 205, 0), (0, 0, 238), (205, 0, 205), (0, 205, 205), (229, 229, 229),
                    (127, 127, 127), (255, 0, 0), (0, 255, 0), (255, 255, 0), (92, 92, 255), (255, 0, 255), (0, 255, 255), (255, 255, 255)]
            return base[n]
        if n < 232:
            n -= 16
            r, g, b = n // 36, (n // 6) % 6, n % 6
            return tuple(0 if v == 0 else 55 + v * 40 for v in (r, g, b))
        v = 8 + (n - 232) * 10
        return (v, v, v)

    def cwidth(ch_):
        if unicodedata.combining(ch_):
            return 0
        return 2 if unicodedata.east_asian_width(ch_) in ("W", "F") else 1

    lines = ansi_text.split("\n")
    for y, line in enumerate(lines[:rows]):
        fg, bg, boldf, inv, ul = (229, 229, 229), (0, 0, 0), False, False, False
        x = 0
        i = 0
        while i < len(line):
            m = re.match(r"\x1b\[([0-9;]*)m", line[i:])
            if m:
                params = [int(p) if p else 0 for p in m.group(1).split(";")] or [0]
                j = 0
                while j < len(params):
                    p = params[j]
                    if p == 0:
                        fg, bg, boldf, inv, ul = (229, 229, 229), (0, 0, 0), False, False, False
                    elif p == 1:
                        boldf = True
                    elif p == 4:
                        ul = True
                    elif p == 7:
                        inv = True
                    elif p == 22:
                        boldf = False
                    elif p == 24:
                        ul = False
                    elif p == 27:
                        inv = False
                    elif p == 39:
                        fg = (229, 229, 229)
                    elif p == 49:
                        bg = (0, 0, 0)
                    elif 30 <= p <= 37:
                        fg = c256(p - 30)
                    elif 90 <= p <= 97:
                        fg = c256(p - 90 + 8)
                    elif 40 <= p <= 47:
                        bg = c256(p - 40)
                    elif 100 <= p <= 107:
                        bg = c256(p - 100 + 8)
                    elif p in (38, 48) and j + 1 < len(params):
                        if params[j + 1] == 2 and j + 4 < len(params):
                            col = tuple(params[j + 2:j + 5])
                            j += 4
                        elif params[j + 1] == 5 and j + 2 < len(params):
                            col = c256(params[j + 2])
                            j += 2
                        else:
                            col = None
                        if col is not None:
                            if p == 38:
                                fg = col
                            else:
                                bg = col
                    j += 1
                i += m.end()
                continue
            if line[i] == "\x1b":
                i += 1
                continue
            ch_ = line[i]
            w = cwidth(ch_)
            f, b = (bg, fg) if inv else (fg, bg)
            if x < cols:
                d.rectangle([x * cw, y * ch, (x + max(w, 1)) * cw - 1, (y + 1) * ch - 1], fill=b)
                if ch_ != " ":
                    d.text((x * cw, y * ch), ch_, font=fontb if boldf else font, fill=f)
                if ul:
                    d.line([x * cw, (y + 1) * ch - 2, (x + max(w, 1)) * cw - 1, (y + 1) * ch - 2], fill=f)
            x += w
            i += 1
        # tmux trims trailing blank cells: extend the last background to the edge
        if x < cols:
            d.rectangle([x * cw, y * ch, cols * cw - 1, (y + 1) * ch - 1], fill=bg)
    img.save(out_path)


def main():
    ap = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("-s", "--size", default="130x42", help="COLSxROWS (default 130x42)")
    ap.add_argument("-k", "--keys", default="", help="space-separated tmux key names sent before the shot")
    ap.add_argument("-m", "--mouse", action="append", default=[], help="mouse step, e.g. 'click 40 12' (repeatable)")
    ap.add_argument("--script", default="", help="';'-separated steps: keys, 'mouse click X Y', 'wait S', 'shot PATH', 'text'")
    ap.add_argument("-o", "--out", default="/tmp/shot.png")
    ap.add_argument("--text", action="store_true", help="print plain-text capture instead of writing a PNG")
    ap.add_argument("--startup", type=float, default=0.8, help="seconds to wait after launch")
    ap.add_argument("cmd", nargs=argparse.REMAINDER, help="-- binary and args")
    a = ap.parse_args()
    cmd = a.cmd[1:] if a.cmd and a.cmd[0] == "--" else a.cmd
    if not cmd:
        ap.error("need -- <binary> [args]")
    cols, rows = (int(v) for v in a.size.lower().split("x"))
    tm = Tmux(cmd, cols, rows)
    try:
        time.sleep(a.startup)
        if a.keys:
            tm.keys(*a.keys.split())
        for step in a.mouse:
            kind, x, y = step.split()
            tm.mouse(kind, int(x), int(y))
        if a.script:
            for step in a.script.split(";"):
                step = step.strip()
                if not step:
                    continue
                parts = step.split()
                if parts[0] == "wait":
                    time.sleep(float(parts[1]))
                elif parts[0] == "shot":
                    render(tm.capture(), cols, rows, parts[1])
                    print("wrote", parts[1])
                elif parts[0] == "text":
                    print(tm.capture(ansi=False))
                elif parts[0] == "mouse":
                    tm.mouse(parts[1], int(parts[2]), int(parts[3]))
                else:
                    tm.keys(*parts)
        time.sleep(0.2)
        if not tm.alive():
            print("PROCESS DIED", file=sys.stderr)
        if a.text:
            print(tm.capture(ansi=False))
        elif not a.script:
            render(tm.capture(), cols, rows, a.out)
            print("wrote", a.out)
    finally:
        err = tm.close()
        if err.strip():
            print("--- stderr ---\n" + err[-2000:], file=sys.stderr)


if __name__ == "__main__":
    main()

File: xtask/src/shot.rs
//! Headless TUI screenshots: run a binary in a private tmux server, drive it with keys and
//! mouse, capture with colours and rasterize to PNG (or print plain text).
//!
//! Key names are tmux send-keys names (Tab, BTab, Enter, Escape, Up, Down, Left, Right,
//! Space, PageUp, PageDown, Home, End, F1.., C-p for ctrl, M-1 for alt, plain characters).
//! `Escape` is always sent alone with a pause, because crossterm reads ESC immediately
//! followed by another key as Alt+key.

use ab_glyph::{Font, FontVec, PxScale, ScaleFont};
use std::process::Command;
use std::time::Duration;
use unicode_width::UnicodeWidthChar;

const FONT_CANDIDATES: [&str; 4] = [
    "/usr/share/fonts/TTF/JetBrainsMonoNerdFontMono-Regular.ttf",
    "/usr/share/fonts/TTF/JetBrainsMono-Regular.ttf",
    "/usr/share/fonts/TTF/DejaVuSansMono.ttf",
    "/usr/share/fonts/truetype/dejavu/DejaVuSansMono.ttf",
];
const FONT_BOLD: [&str; 4] = [
    "/usr/share/fonts/TTF/JetBrainsMonoNerdFontMono-Bold.ttf",
    "/usr/share/fonts/TTF/JetBrainsMono-Bold.ttf",
    "/usr/share/fonts/TTF/DejaVuSansMono-Bold.ttf",
    "/usr/share/fonts/truetype/dejavu/DejaVuSansMono-Bold.ttf",
];

const CELL_W: u32 = 11;
const CELL_H: u32 = 22;
const DEFAULT_FG: [u8; 3] = [229, 229, 229];
const DEFAULT_BG: [u8; 3] = [0, 0, 0];

pub fn run(args: &[String]) -> Result<(), String> {
    let mut size = (130u32, 42u32);
    let mut keys = String::new();
    let mut mouse: Vec<String> = Vec::new();
    let mut out = "/tmp/shot.png".to_string();
    let mut text_only = false;
    let mut startup = 0.8f32;
    let mut cmd: Vec<String> = Vec::new();

    let mut it = args.iter();
    while let Some(a) = it.next() {
        let mut next = || it.next().cloned().ok_or_else(|| format!("{a} needs a value"));
        match a.as_str() {
            "-s" | "--size" => {
                let v = next()?;
                let (c, r) = v
                    .to_lowercase()
                    .split_once('x')
                    .map(|(c, r)| (c.to_string(), r.to_string()))
                    .ok_or("size must look like 130x42")?;
                size = (
                    c.parse().map_err(|_| "bad columns")?,
                    r.parse().map_err(|_| "bad rows")?,
                );
            }
            "-k" | "--keys" => keys = next()?,
            "-m" | "--mouse" => mouse.push(next()?),
            "-o" | "--out" => out = next()?,
            "--text" => text_only = true,
            "--startup" => startup = next()?.parse().map_err(|_| "bad startup seconds")?,
            "--" => {
                cmd.extend(it.cloned());
                break;
            }
            other => return Err(format!("unknown shot option `{other}`")),
        }
    }
    if cmd.is_empty() {
        return Err("shot: need -- <binary> [args]".into());
    }

    let tmux = Tmux::start(&cmd, size.0, size.1)?;
    let result = (|| -> Result<(), String> {
        sleep(startup);
        for k in keys.split_whitespace() {
            tmux.key(k)?;
        }
        for step in &mouse {
            let parts: Vec<&str> = step.split_whitespace().collect();
            let [kind, x, y] = parts[..] else {
                return Err(format!("mouse step must be `kind X Y`, got `{step}`"));
            };
            tmux.mouse(
                kind,
                x.parse().map_err(|_| "bad mouse x")?,
                y.parse().map_err(|_| "bad mouse y")?,
            )?;
        }
        sleep(0.2);
        if !tmux.alive() {
            eprintln!("shot: the process died before the capture");
        }
        if text_only {
            print!("{}", tmux.capture(false)?);
        } else {
            let ansi = tmux.capture(true)?;
            render(&ansi, size.0, size.1, &out)?;
            println!("wrote {out}");
        }
        Ok(())
    })();

    let stderr = tmux.close();
    if !stderr.trim().is_empty() {
        eprintln!("--- stderr ---\n{}", tail(&stderr, 2000));
    }
    result
}

fn sleep(secs: f32) {
    std::thread::sleep(Duration::from_secs_f32(secs));
}

fn tail(s: &str, n: usize) -> &str {
    if s.len() <= n { s } else { &s[s.len() - n..] }
}

struct Tmux {
    socket: String,
    err_path: String,
}

impl Tmux {
    fn start(cmd: &[String], cols: u32, rows: u32) -> Result<Self, String> {
        let socket = format!("shot-{}", std::process::id());
        let err_path = format!("/tmp/shot-{}.err", std::process::id());
        let shell = format!(
            "{} 2>{}",
            cmd.iter().map(|c| quote(c)).collect::<Vec<_>>().join(" "),
            err_path
        );
        let me = Tmux { socket, err_path };

        // CI and agent shells often export NO_COLOR=1 or TERM=dumb. The tmux server, and so
        // the app under test, inherits that and crossterm would strip every colour.
        let status = me
            .cmd()
            .args(["new-session", "-d", "-s", "s", "-x"])
            .arg(cols.to_string())
            .arg("-y")
            .arg(rows.to_string())
            .arg(&shell)
            .env_remove("NO_COLOR")
            .env_remove("TMUX")
            .env_remove("TMUX_PANE")
            .env("TERM", "xterm-256color")
            .env("COLORTERM", "truecolor")
            .status()
            .map_err(|e| format!("shot: cannot run tmux: {e}"))?;
        if !status.success() {
            return Err("shot: tmux new-session failed".into());
        }
        let _ = me.cmd().args(["set", "-g", "status", "off"]).status();
        Ok(me)
    }

    fn cmd(&self) -> Command {
        let mut c = Command::new("tmux");
        c.args(["-L", &self.socket]);
        c
    }

    fn key(&self, k: &str) -> Result<(), String> {
        self.cmd()
            .args(["send-keys", "-t", "s", k])
            .status()
            .map_err(|e| format!("shot: send-keys: {e}"))?;
        // ESC followed immediately by another key is parsed as Alt+key.
        sleep(if k == "Escape" || k == "Esc" { 0.35 } else { 0.0 });
        sleep(0.25);
        Ok(())
    }

    fn mouse(&self, kind: &str, x: u32, y: u32) -> Result<(), String> {
        if kind == "click" {
            self.mouse("down", x, y)?;
            return self.mouse("up", x, y);
        }
        let (button, suffix) = match kind {
            "down" => (0, 'M'),
            "up" => (0, 'm'),
            "drag" => (32, 'M'),
            "move" => (35, 'M'),
            "wheelup" => (64, 'M'),
            "wheeldown" => (65, 'M'),
            other => return Err(format!("unknown mouse step `{other}`")),
        };
        let seq = format!("\x1b[<{button};{};{}{suffix}", x + 1, y + 1);
        self.cmd()
            .args(["send-keys", "-t", "s", "-l", &seq])
            .status()
            .map_err(|e| format!("shot: mouse: {e}"))?;
        sleep(0.15);
        Ok(())
    }

    fn capture(&self, ansi: bool) -> Result<String, String> {
        let mut c = self.cmd();
        c.args(["capture-pane", "-t", "s", "-p"]);
        if ansi {
            c.arg("-e");
        }
        let out = c.output().map_err(|e| format!("shot: capture-pane: {e}"))?;
        Ok(String::from_utf8_lossy(&out.stdout).into_owned())
    }

    fn alive(&self) -> bool {
        let out = self
            .cmd()
            .args(["list-panes", "-t", "s", "-F", "#{pane_dead}"])
            .output();
        matches!(out, Ok(o) if o.status.success() && String::from_utf8_lossy(&o.stdout).trim() == "0")
    }

    fn close(&self) -> String {
        let _ = self.cmd().arg("kill-server").output();
        let err = std::fs::read_to_string(&self.err_path).unwrap_or_default();
        let _ = std::fs::remove_file(&self.err_path);
        err
    }
}

fn quote(s: &str) -> String {
    format!("'{}'", s.replace('\'', "'\\''"))
}

fn load_font(candidates: &[&str]) -> Option<FontVec> {
    candidates
        .iter()
        .find_map(|p| std::fs::read(p).ok())
        .and_then(|bytes| FontVec::try_from_vec(bytes).ok())
}

/// xterm's 256-colour cube.
fn color256(n: u8) -> [u8; 3] {
    const BASE: [[u8; 3]; 16] = [
        [0, 0, 0],
        [205, 0, 0],
        [0, 205, 0],
        [205, 205, 0],
        [0, 0, 238],
        [205, 0, 205],
        [0, 205, 205],
        [229, 229, 229],
        [127, 127, 127],
        [255, 0, 0],
        [0, 255, 0],
        [255, 255, 0],
        [92, 92, 255],
        [255, 0, 255],
        [0, 255, 255],
        [255, 255, 255],
    ];
    match n {
        0..=15 => BASE[n as usize],
        16..=231 => {
            let n = n - 16;
            let step = |v: u8| if v == 0 { 0 } else { 55 + v * 40 };
            [step(n / 36), step((n / 6) % 6), step(n % 6)]
        }
        _ => {
            let v = 8 + (n - 232) * 10;
            [v, v, v]
        }
    }
}

struct Pen {
    fg: [u8; 3],
    bg: [u8; 3],
    bold: bool,
    inverse: bool,
    underline: bool,
}

impl Pen {
    fn reset() -> Self {
        Pen { fg: DEFAULT_FG, bg: DEFAULT_BG, bold: false, inverse: false, underline: false }
    }

    fn apply(&mut self, params: &[u16]) {
        let mut i = 0;
        while i < params.len() {
            match params[i] {
                0 => *self = Pen::reset(),
                1 => self.bold = true,
                4 => self.underline = true,
                7 => self.inverse = true,
                22 => self.bold = false,
                24 => self.underline = false,
                27 => self.inverse = false,
                39 => self.fg = DEFAULT_FG,
                49 => self.bg = DEFAULT_BG,
                p @ 30..=37 => self.fg = color256((p - 30) as u8),
                p @ 90..=97 => self.fg = color256((p - 90 + 8) as u8),
                p @ 40..=47 => self.bg = color256((p - 40) as u8),
                p @ 100..=107 => self.bg = color256((p - 100 + 8) as u8),
                p @ (38 | 48) if i + 1 < params.len() => {
                    let colour = match params[i + 1] {
                        2 if i + 4 < params.len() => {
                            let c = [params[i + 2] as u8, params[i + 3] as u8, params[i + 4] as u8];
                            i += 4;
                            Some(c)
                        }
                        5 if i + 2 < params.len() => {
                            let c = color256(params[i + 2] as u8);
                            i += 2;
                            Some(c)
                        }
                        _ => None,
                    };
                    if let Some(c) = colour {
                        if p == 38 { self.fg = c } else { self.bg = c }
                    }
                }
                _ => {}
            }
            i += 1;
        }
    }
}

fn render(ansi: &str, cols: u32, rows: u32, out: &str) -> Result<(), String> {
    let regular = load_font(&FONT_CANDIDATES)
        .ok_or("shot: no monospace TTF found; install JetBrains Mono or DejaVu Sans Mono")?;
    let bold = load_font(&FONT_BOLD);
    // Box-drawing glyphs tile only if one advance is exactly one cell. Stretch the face
    // horizontally to CELL_W instead of leaving the font's natural advance, which is
    // narrower and turns a run of `─` into a dashed line.
    let natural = regular.as_scaled(PxScale::from(18.0));
    let advance = natural.h_advance(regular.glyph_id('M'));
    let scale = PxScale { x: 18.0 * CELL_W as f32 / advance, y: 18.0 };

    let mut img = image::RgbImage::from_pixel(cols * CELL_W, rows * CELL_H, image::Rgb(DEFAULT_BG));

    for (y, line) in ansi.split('\n').take(rows as usize).enumerate() {
        let y = y as u32;
        let mut pen = Pen::reset();
        let mut x = 0u32;
        let mut chars = line.chars().peekable();

        while let Some(c) = chars.next() {
            if c == '\x1b' {
                if chars.peek() == Some(&'[') {
                    chars.next();
                    let mut raw = String::new();
                    for c in chars.by_ref() {
                        if c.is_ascii_alphabetic() {
                            if c == 'm' {
                                let params: Vec<u16> = if raw.is_empty() {
                                    vec![0]
                                } else {
                                    raw.split(';').map(|p| p.parse().unwrap_or(0)).collect()
                                };
                                pen.apply(&params);
                            }
                            break;
                        }
                        raw.push(c);
                    }
                }
                continue;
            }

            let w = c.width().unwrap_or(1).max(1) as u32;
            let (fg, bg) = if pen.inverse { (pen.bg, pen.fg) } else { (pen.fg, pen.bg) };
            if x < cols {
                fill_cells(&mut img, x, y, w, bg);
                if c != ' ' {
                    let font = if pen.bold { bold.as_ref().unwrap_or(&regular) } else { &regular };
                    draw_glyph(&mut img, font, scale, c, x, y, fg);
                }
                if pen.underline {
                    let uy = (y + 1) * CELL_H - 2;
                    for px in x * CELL_W..((x + w) * CELL_W).min(img.width()) {
                        img.put_pixel(px, uy.min(img.height() - 1), image::Rgb(fg));
                    }
                }
            }
            x += w;
        }
        // tmux trims trailing blank cells: extend the last background to the edge.
        if x < cols {
            fill_cells(&mut img, x, y, cols - x, pen.bg);
        }
    }

    img.save(out).map_err(|e| format!("shot: writing {out}: {e}"))
}

fn fill_cells(img: &mut image::RgbImage, x: u32, y: u32, w: u32, colour: [u8; 3]) {
    let (x0, y0) = (x * CELL_W, y * CELL_H);
    for py in y0..(y0 + CELL_H).min(img.height()) {
        for px in x0..(x0 + w * CELL_W).min(img.width()) {
            img.put_pixel(px, py, image::Rgb(colour));
        }
    }
}

fn draw_glyph(
    img: &mut image::RgbImage,
    font: &FontVec,
    scale: PxScale,
    c: char,
    cell_x: u32,
    cell_y: u32,
    fg: [u8; 3],
) {
    let scaled = font.as_scaled(scale);
    let glyph = font.glyph_id(c).with_scale(scale);
    let Some(outline) = font.outline_glyph(glyph) else { return };
    let bounds = outline.px_bounds();
    // Baseline placement inside the cell, matching the ascent of an 18px face in a 22px cell.
    let origin_x = (cell_x * CELL_W) as f32;
    let origin_y = (cell_y * CELL_H) as f32 + scaled.ascent();
    outline.draw(|gx, gy, coverage| {
        if coverage <= 0.01 {
            return;
        }
        let px = (origin_x + bounds.min.x + gx as f32).round();
        let py = (origin_y + bounds.min.y + gy as f32).round();
        if px < 0.0 || py < 0.0 {
            return;
        }
        let (px, py) = (px as u32, py as u32);
        if px >= img.width() || py >= img.height() {
            return;
        }
        let bg = img.get_pixel(px, py).0;
        let a = coverage.clamp(0.0, 1.0);
        let blend = |f: u8, b: u8| (f as f32 * a + b as f32 * (1.0 - a)).round() as u8;
        img.put_pixel(
            px,
            py,
            image::Rgb([blend(fg[0], bg[0]), blend(fg[1], bg[1]), blend(fg[2], bg[2])]),
        );
    });
}
